### PR TITLE
Refactor Observability plugin to use signals for reacting to location state changes

### PR DESCRIPTION
**Pull request by colossatr0n**
Previous to this PR, the Observability plugin would track the location state during draw loops. This is error prone as the developer must ensure to reset its state in the appropriate places.

What this PR does is let StelLocationMgr be the sole owner of the state by using signals and slots to react to the location change event. This is the first PR in a series of PRs that are derived from a branch I'm working on. I'll try to keep these PRs as focused as possible.

There are A LOT of whitespace/formatting changes. I used the requested formatting configuration, and I guess it hasn't been run on those files for a long time.

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

The changes in this PR are relatively simple. No intentional changes have been made to the calculation logic.

Testing consisted of running Stellarium, enabling the Observability plugin, clicking an object, and ensuring that a report showed up in the bottom left hand side of the screen.

Application logs were also checked to ensure that connections were made and closed at the appropriate times. They were also checked to see if new errors were introduced.

**Test Configuration**:
* Operating system: macOS Catalina v10.15.7
* Graphics Card: Intel Iris 1536 MB

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: plugins/Observability/src/Observability.cpp [hunk 1]
```diff
@@ -44,1804 +44,1736 @@
...
+   info.id            = "Observability";
+   info.displayedName = N_("Observability Analysis");
+   info.authors       = "Ivan Marti-Vidal (Onsala Space Observatory)"; // non-translatable field
+   info.contact       = "i.martividal@gmail.com";
+   info.description =
+     N_("Displays an analysis of a selected object's observability (rise, set, and transit times) for the current "
+	"date, as well as when it is observable through the year. An object is assumed to be observable if it is above "
+	"the horizon during a fraction of the night. Also included are the dates of the largest separation from the "
+	"Sun and acronychal and cosmical rising and setting. (Explanations are provided in the 'About' tab of the "
+	"plugin's configuration window.)");
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:72]: Auto formatting.

File: plugins/Observability/src/Observability.cpp [hunk 2]
```diff
@@ -44,1804 +44,1736 @@
...
-const double Observability::Rad2Hr = 12./M_PI;           // Convert hours into radians
-const double Observability::UA =  AU;                    // 1.4958e+8;         // Astronomical Unit in Km. ==> HAS BEEN DEFINED IN StelUtils.hpp!
-const double Observability::TFrac = 0.9972677595628414;  // Convert sidereal time into Solar time
-const double Observability::halfpi = M_PI_2;         //  1.57079632675; // pi/2
-const double Observability::MoonT = 29.530588;           // Moon synodic period (used as first estimate of Full Moon). ==> FIND MORE DEC. PLACES!
-const double Observability::RefFullMoon = 2451564.696; // Reference Julian date of a Full Moon.
+const double Observability::Rad2Deg = M_180_PI;   // Convert degrees into radians
+const double Observability::Rad2Hr  = 12. / M_PI; // Convert hours into radians
+const double Observability::UA =
+  AU; // 1.4958e+8;         // Astronomical Unit in Km. ==> HAS BEEN DEFINED IN StelUtils.hpp!
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:82]: Auto formatting.

File: plugins/Observability/src/Observability.cpp [hunk 3]
```diff
@@ -44,1804 +44,1736 @@
...
+const double Observability::TFrac  = 0.9972677595628414; // Convert sidereal time into Solar time
+const double Observability::halfpi = M_PI_2;             //  1.57079632675; // pi/2
+const double Observability::MoonT =
+  29.530588; // Moon synodic period (used as first estimate of Full Moon). ==> FIND MORE DEC. PLACES!
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:86]: Auto formatting.

File: plugins/Observability/src/Observability.cpp [hunk 4]
```diff
@@ -44,1804 +44,1736 @@
...
-	connect(StelApp::getInstance().getCore(), SIGNAL(configurationDataSaved()), this, SLOT(saveConfiguration()));
+   loadConfiguration();
+
+   addAction("actionShow_Observability", N_("Observability"), N_("Observability"), "flagShowReport");
+   addAction("actionShow_Observability_dialog",
+	     N_("Observability"),
+	     N_("Show settings dialog"),
+	     configDialog,
+	     "visible",
+	     ""); // Allow assign shortkey
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:220]: Auto formatting.

File: plugins/Observability/src/Observability.cpp [hunk 5]
```diff
@@ -44,1804 +44,1736 @@
...
+
+   StelGui * gui = dynamic_cast<StelGui *>(StelApp::getInstance().getGui());
+   if (gui != Q_NULLPTR) {
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:223]: Auto formatting.

File: plugins/Observability/src/Observability.cpp [hunk 6]
```diff
@@ -44,1804 +44,1736 @@
...
+			      QPixmap(":/observability/bt_observab_off.png"),
+			      QPixmap(":/graphicGui/miscGlow32x32.png"),
+			      "actionShow_Observability",
+			      false,
+			      "actionShow_Observability_dialog");
+      gui->getButtonBar()->addButton(button, "065-pluginsGroup");
+   }
+
+   updateMessageText();
+   connect(&StelApp::getInstance(), SIGNAL(languageChanged()), this, SLOT(updateMessageText()));
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:235]: Auto formatting.

**Review by alex-w — CHANGES_REQUESTED** [review]
See the obvious issue:
stellarium-009

Plus the information is not translated when language is changed.

**Comment by colossatr0n** [comment]
@alex-w Can you tell me what steps you took for report to not show information? I set up stellarium with the same parameters, but the data is available in my setup:
image

**Comment by alex-w** [comment]
1. Run Stellarium and enable the plugin
2. Select any object
3. Open Location dialog and change the location (you can do it twice or more times)
4. Press "Return to default location"

**Comment by colossatr0n** [comment]
@alex-w Both the language issue and report issue are fixed. Can you verify that it works for you as well?

**Comment by worachate001** [comment]
@colossatr0n It's stop working after you click the Moon.
Select some stars -> select the Moon -> select a star.

File: plugins/Observability/src/Observability.cpp [hunk 7]
```diff
@@ -44,1804 +44,1749 @@
...
-// Only execute plugin if we are on Earth.
-	if (core->getCurrentLocation().planetName != "Earth")
-		return;
-
-// Set the painter:
-	StelPainter painter(core->getProjection2d());
-	painter.setColor(fontColor[0],fontColor[1],fontColor[2],1.f);
-	font.setPixelSize(fontSize);
-	painter.setFont(font);
-
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:252]: Moved this into a function.

File: plugins/Observability/src/Observability.cpp [hunk 8]
```diff
@@ -44,1804 +44,1749 @@
...
-	else
-	{
-		locChanged = true;
-		mylat = currlat; mylon = currlon;
-		double temp1 = currheight*std::cos(currlat);
-		ObserverLoc[0] = temp1*std::cos(currlon);
-		ObserverLoc[1] = temp1*std::sin(currlon);
-		ObserverLoc[2] = currheight*std::sin(currlat);
-	}
-
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:304]: This is encapsulated in `onLocationChanged`.

**Comment by colossatr0n** [comment]
@worachate001 Just pushed a fix. Thanks for finding that one.

**Comment by worachate001** [comment]
> @worachate001 Just pushed a fix. Thanks for finding that one.

It works as expected now, thank you.

**Comment by alex-w** [comment]
Important note: our .clang* file are not completed and rules didn't fully follow our coding style at the moment

**Comment by colossatr0n** [comment]
@alex-w Is there a TODO for me regarding your last comment?

**Comment by colossatr0n** [comment]
> > @alex-w Is there a TODO for me regarding your last comment?
> 
> Probably you should re-check the coding style in this PR (did you used our settings file for QtCreator?)

Yes. I used the provided XML file. Maybe I did it incorrectly? I selected all the code > Edit > Advanced > Auto format selection.

**Comment by alex-w** [comment]
> Yes. I used the provided XML file. Maybe I did it incorrectly? I selected all the code > Edit > Advanced > Auto format selection.

Good question! I've tried apply auto format in QtCreator 8 (macOS) with active "Stellarium" C++ style and he is formatted the code as described in Qt style. :-/

File: plugins/Observability/src/Observability.cpp [hunk 9]
```diff
@@ -243,19 +252,10 @@ void Observability::draw(StelCore* core)
...
-// Set the painter:
-	StelPainter painter(core->getProjection2d());
-	painter.setColor(fontColor[0],fontColor[1],fontColor[2],1.f);
-	font.setPixelSize(fontSize);
-	painter.setFont(font);
-
-// Get current date, location, and check if there is something selected.
-	double currlat = static_cast<double>(core->getCurrentLocation().latitude)/Rad2Deg;
-	double currlon = static_cast<double>(core->getCurrentLocation().longitude)/Rad2Deg;
-	double currheight = (6371.+(core->getCurrentLocation().altitude)/1000.)/UA;
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:255]: Moved into `onLocationChanged`

File: plugins/Observability/src/Observability.cpp [hunk 10]
```diff
@@ -644,212 +616,19 @@ void Observability::draw(StelCore* core)
...
-					{
-						tempPhs = Lambda(objectRA[i], objectDec[i],
-						                 sunRA[i], sunDec[i]);
-						if (tempPhs > deltaPhs)
-						{
-							selday = i;
-							deltaPhs = tempPhs;
-						}
-					}
+      // Determine source observability (only if something changed):
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:619]: The hunk below this comment was turned into the `getObjectObservability()` method.

File: plugins/Observability/src/Observability.cpp [hunk 11]
```diff
@@ -644,212 +616,19 @@ void Observability::draw(StelCore* core)
...
+      if (souChanged || yearChanged) {
+            getObjectObservability();
+      } // Comes from  "souChanged || ..."
+   }    // Comes from the "else" with "!isMoon"
```
Comment (colossatr0n) [inline plugins/Observability/src/Observability.cpp:624]: The hunk below this comment was converted into the `renderResults` method.

**Comment by alex-w** [comment]
@gzotti please review the PR (and please merge it if this is OK for you)

File: plugins/Observability/src/Observability.cpp [hunk 12]
```diff
@@ -208,23 +207,33 @@ void Observability::init()
...
+      gui->getButtonBar()->addButton(button, "065-pluginsGroup");
+   }
+
+   updateMessageText();
+   connect(&StelApp::getInstance(), &StelApp::languageChanged, this, &Observability::onLanguageChanged);
+   connect(StelApp::getInstance().getCore(), SIGNAL(configurationDataSaved()), this, SLOT(saveConfiguration()));
+
+   connect(this, &Observability::flagReportVisibilityChanged, this, [&](bool enabled) {
+      if (enabled) {
+        qDebug() << "[Observability] Creating plugin status dependent connections.";
```
Comment (gzotti) [inline plugins/Observability/src/Observability.cpp:228]: You can comment these two debug messages away in non-debug builds. Either #ifndef NDEBUG or just comment away when they are working.

**Comment by gzotti** [comment]
@colossatr0n please rebase this PR from current master. The online diff tool does not work properly. I don't know what the conflicts are, but they should be minimal.

**Comment by colossatr0n** [comment]
Pushed and cleaned up the messed up formatting. I need to retest, but am having troubles building for some reason. I'm getting build issues saying that the QtScript file can't be found. This was before rebasing, and afterward as well. Building master to see if I can at least get that to build.

**Comment by gzotti** [comment]
QtScript is a deprecated part of Qt5. When installing Qt5 libraries, you must tick a separate box to install it. When you are building with Qt6, you must for now configure  ENABLE_SCRIPTING=false in cmake. I am in the process of re-enabling scripting in Qt6 with the newer scripting engine.

**Comment by colossatr0n** [comment]
Finished testing. Ready to be merged/reviewed.

File: plugins/Observability/src/Observability.cpp [hunk 13]
```diff
@@ -1838,6 +1613,273 @@ void Observability::showReport(bool b)
...
+//	{
+//		//renderer->drawText(TextParams(xLine, yLine,q_("TODAY:")));
+//		painter.drawText(xLine, yLine, msgToday);
+//		painter.drawText(xLine + fontSize, yLine - lineSpacing, RS2);
+//		painter.drawText(xLine + fontSize, yLine - lineSpacing*2, RS1);
+//		painter.drawText(xLine + fontSize, yLine - lineSpacing*3, Cul);
+//		yLine -= groupSpacing;
+//	}
+
+	if ((isMoon(objectSelection) && show_FullMoon) || (!isSun(objectSelection) && !isMoon(objectSelection) && shouldShowYear()))
```
Comment (gzotti) [inline plugins/Observability/src/Observability.cpp:1847]: objectSelection seems to be unknown. Did you push the latest version?

File: plugins/Observability/src/Observability.cpp [hunk 14]
```diff
@@ -44,10 +44,10 @@
 #include "StelObject.hpp"
 #include "StelObjectMgr.hpp"
 #include "StelObserver.hpp"
```
Comment (gzotti) [inline plugins/Observability/src/Observability.cpp:46]: My QtCreator suggests that following headers may be removed:
StarMgr, StelFader, StelFileMgr, StelIniParser, StelObserver, StelSkyDrawer, ZoneArray.

It is possible that omission of one header requires inclusion of a "lighter" one. Just make sure it compiles.

Comment (colossatr0n) [reply plugins/Observability/src/Observability.cpp:46]: I'll remove the unnecessary ones and see if it compiles.

File: plugins/Observability/src/Observability.cpp (hunk 13 above)
Comment (colossatr0n) [reply plugins/Observability/src/Observability.cpp:1849]: Oh, whoops. Forgot to push that. Pushed.

**Comment by colossatr0n** [comment]
This is ready to be reviewed/merged.

**Comment by gzotti** [comment]
@colossatr0n Thank you!

**Pull request merged**